element_calcium_imaging/scan.py

```diff
@@ -613,3 +607,142 @@ def estimate_nd2_scan_duration(nd2_scan_obj):
+
+        Attributes:
+            ScanInfo.Field (foreign key): Primary key from ScanInfo.Field.
+            channel (int): Channel index.
+            min_intensity (int): Minimum value in movie.
+            max_intensity (int): Maximum value in movie.
+            quantal_size (float): Quantal size (variance/mean intensity slope), gain.
+            zero_level (int): Level corresponding to zero (computed from variance dependence).
+            quantal_frame (longblob): Average frame expressed in quanta.
+
+        """
+
+        definition = """ 
+        -> master
+        channel         : int      # Channel index
+        ---
+        min_intensity   : int      # Minimum value in movie.
+        max_intensity   : int      # Maximum value in movie.
+        quantal_size    : float    # Quantal size (variance/mean intensity slope), gain.
+        zero_level      : int      # Level corresponding to zero (computed from variance dependence)
+        quantal_frame   : longblob # Average frame expressed in quanta.
+        """
+
+    def make(self, key):
+        acq_software, nchannels = (Scan * ScanInfo & key).fetch1(
+            "acq_software", "nchannels"
+        )
+
+        self.insert1(key)
+
+        for channel in range(nchannels):
+            if acq_software == "ScanImage":
+                import scanreader
+
+                movie = scanreader.read_scan(get_scan_image_files(key))[
+                    key["field_idx"], :, :, channel, :
+                ].transpose(2, 0, 1)
+            elif acq_software == "Scanbox":
+                import sbxreader
+
+                movie = sbxreader(get_scan_box_files(key))[
+                    :, key["field_idx"], channel, :, :
+                ]
+            elif acq_software == "NIS":
+                import nd2
+
+                nd2_file = nd2.ND2File(get_nd2_files(key)[0])
+
+                nd2_dims = {k: i for i, k in enumerate(nd2_file.sizes)}
+
+                valid_dimensions = set("TZCYX")
+                assert valid_dimensions == set(
+                    nd2_dims
+                ), f"Unknown or missing dimension in {nd2_dims}"
+                movie = nd2_file.asarray().transpose(
+                    [nd2_dims[x] for x in valid_dimensions]
+                )
+
+                for i, dim in enumerate("TZC"):
+                    if dim not in nd2_dims:
+                        movie = np.expand_dims(movie, i)
+
+                movie = movie[:, key["field_idx"], channel, :, :]
+
+            self.FrameMetrics.insert1(
+                dict(
+                    key,
+                    channel=channel,
+                    min_intensity=movie.min(axis=(1, 2)),
+                    mean_intensity=movie.mean(axis=(1, 2)),
+                    max_intensity=movie.max(axis=(1, 2)),
+                    contrast=np.diff(
+                        np.percentile(movie, [1, 99], axis=(1, 2)), axis=0
+                    ).squeeze(),
```

Comment:
```suggestion
                    contrast=np.percentile(movie, 99) - np.percentile(movie, 1)
```

Reply:
This suggestion is not correct. Definition of contrast is per frame in this table.

Reply:
ok, updated